includes/Utilities/Plugin_Request_Utility.php

```diff
@@ -182,6 +182,91 @@ public static function get_files_to_ignore() {
 		return $files_to_ignore;
 	}
 
+	/**
+	 * Returns the plugin basename after downloading and installing the plugin.
+	 *
+	 * @since 1.1.0
+	 *
+	 * @param string $plugin_url The URL of the plugin to download.
+	 * @return string The plugin basename after downloading and installing the plugin.
+	 *
+	 * @throws Exception Thrown if an invalid URL given or zip could be extracted properly.
+	 */
+	public static function download_plugin( $plugin_url ) {
+		$response = wp_remote_get( $plugin_url );
+
+		if ( is_wp_error( $response ) || 200 !== wp_remote_retrieve_response_code( $response ) ) {
+			throw new Exception(
+				__( 'Downloading the zip file failed.', 'plugin-check' )
+			);
+		}
+
+		// Prevents URL from wporg.
+		if ( false !== strpos( $plugin_url, '#wporgapi:' ) ) {
+			$plugin_info_url = substr( $plugin_url, strpos( $plugin_url, '#wporgapi:' ) );
+			$plugin_info_url = str_replace( '#wporgapi:', '', $plugin_info_url );
+			$plugin_url      = substr( $plugin_url, 0, strpos( $plugin_url, '#' ) );
+		}
+
+		$basename = basename( $plugin_url );
+
+		// Create the name of the file and the declare the directory and path.
+		$plugin_check_dir = self::get_upload_dir();
+
+		$file_path    = $plugin_check_dir . $basename;
+		$file_process = fopen( $file_path, 'w' );
+		fwrite( $file_process, $response['body'] );
+		fclose( $file_process );
+
+		$temp_dir = $plugin_check_dir . strtotime( 'now' ) . '/';
+
+		// Unzip file.
+		require_once ABSPATH . 'wp-admin/includes/file.php';
+
+		WP_Filesystem();
```

Comment:
This doesn't seem to be used at all.

Either we should use the filesystem API or `fwrite` & co, but not both.

Reply:
I think `unzip_file()` requires this call. https://developer.wordpress.org/reference/functions/unzip_file/